### PR TITLE
fix for out-of-bounds index in `HitPattern::getHitPatternByAbsoluteIndex`

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR contains a minor technical fix to `HitPattern::getHitPatternByAbsoluteIndex`.

It was prompted by an error detected in [last week's UBSAN IB (`step4` of wf `4.22`)](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ubsan_logs/CMSSW_12_2_X_2021-11-05-2300/logs/64/64af2f39a52f88113283751b65fc7259/log).

The error was coming from `secondWord==57` (out of bounds) when `position==75`.

Merely technical. No changes expected.

#### PR validation:

Tested wf `4.22` with a UBSAN IB.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR:

N/A

### CODE REVIEW

File: DataFormats/TrackReco/src/HitPattern.cc [hunk 1]
```diff
@@ -334,6 +334,8 @@ uint16_t HitPattern::getHitPatternByAbsoluteIndex(int position) const {
   } else {
     uint8_t firstWordBits = HIT_LENGTH - secondWordBits;
     uint16_t firstWordBlock = hitPattern[secondWord - 1] >> (16 - firstWordBits);
+    if (!secondWordBits)
```
Comment (VinInn) [inline DataFormats/TrackReco/src/HitPattern.cc:337]: what do you mean with ```!``` of a uint8?

Comment (missirol) [reply DataFormats/TrackReco/src/HitPattern.cc:337]: Just wanted this to be `true` if all bits were zero.
Testing locally (outside CMSSW) it seemed to work, but I guess I'm missing something?

Comment (slava77) [reply DataFormats/TrackReco/src/HitPattern.cc:337]: `!= 0` ?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Looking at this code, it's not clear why lines 3-4 would do anything different than line 2 if `secondWordBits==0`. In other words, could you please explain a bit how this fixes the out of bounds issue reported by UBSAN?
```C++
1:    if (secondWordBits == 0)
2:     return firstWordBlock;
3:    uint16_t secondWordBlock = hitPattern[secondWord] & ((1 << secondWordBits) - 1); //(1<<0)-1 gives secondWordBlock==0
4:    uint16_t myResult = firstWordBlock + (secondWordBlock << firstWordBits); //0 << firstWordBits gives 0
```

**Comment by missirol** [comment]
The difference is that with this change, when `secondWordBits == 0`, `hitPattern` is not accessed in L3; this is what fixes the out-of-bounds issue reported in the UBSAN IB for the edge case `secondWord == 57` (`hitPattern` has size `57`, so `57` is out of bounds). The return value of the relevant function is unchanged by this PR, because (even pre-PR) if `secondWordBits==0`, then `myResult` equals `firstWordBlock`.

**Comment by jpata** [comment]
ah, I see, thanks for the explanation!

**Comment by jpata** [comment]
+reconstruction
- technical UBSAN fix
- no reco changes

**Comment by perrotta** [comment]
+1

**Pull request merged**